urlbf4fd38f68_rduerig_StrategyBDD_tgz-pJ8-com_strategy_util_TestMainJ8/src/com/strategy/prototype/board/BoardPrototype.java

```diff
@@ -0,0 +1,212 @@
+			throw new IllegalArgumentException("Given board was empty!");
+		}
+
+		rows = board.length;
+		columns = board[0].length;
+		fields = new Field[rows][columns];
+		positions = new HashSet<Position>();
+		init(board);
+	}
+
+	public static Board createInstance(int[][] board) {
+		return new BoardPrototype(board);
+	}
+
+	// ************************************************************************
+
+	@Override
+	public Field getField(int row, int col) {
+		if (!isInRange(row, col)) {
+			return null;
+		}
+		return fields[row][col];
+	}
+
+	@Override
+	public Field getField(RowConstant coord, Integer coordNumber) {
+		return null;
+	}
+
+	@Override
+	public Field getField(int index) {
+		return getField(index / getRows(), index % getRows());
+	}
+
+	@Override
+	public void setField(Field newField) {
+		int row = newField.getPosition().getRow();
+		int col = newField.getPosition().getCol();
+		if (isInRange(row, col)) {
+			fields[row][col] = newField;
+		}
+	}
+
+	@Override
+	public int getColumns() {
+		return columns;
+	}
+
+	@Override
+	public int getRows() {
+		return rows;
+	}
+
+	@Override
+	public int getBoardSize() {
+		return rows;
+	}
+
+	@Override
+	public Collection<Position> getPositions() {
+		return positions;
+	}
+
+	@Override
+	public boolean isValidField(int row, int col) {
+		return isValidField(PositionSquare.get(row, col));
+	}
+
+	@Override
+	public boolean isValidField(int index) {
+		return null != getField(index);
+	}
+
+	@Override
+	public boolean isValidField(Position p) {
+		return positions.contains(p);
+	}
+
+	@Override
+	public boolean isEmptyField(int index) {
+		return getField(index) instanceof EmptyField;
+	}
+
+	@Override
+	public String toString() {
+		StringBuilder sb = new StringBuilder();
+		for (int i = 0; i < rows; i++) {
+			for (int j = 0; j < columns; j++) {
+				sb.append(fields[i][j] + "");
+			}
+			sb.append("\n");
+		}
+		return sb.toString();
+	}
+
+	@Override
+	public String toMarkLastTurnString(Integer index) {
+		StringBuilder sb = new StringBuilder();
+		for (int i = 0; i < rows; i++) {
+			for (int j = 0; j < columns; j++) {
+				Field field = fields[i][j];
+				if (null != index && index == field.getIndex()) {
+					sb.append("(" + field + ")");
+				} else {
+					sb.append(field + "");
+				}
+			}
+			sb.append("\n");
+		}
+		return sb.toString();
+	}
+
+	@Override
+	public String toIndexString() {
+		StringBuilder sb = new StringBuilder();
+		for (int i = 0; i < rows; i++) {
+			for (int j = 0; j < columns; j++) {
+				sb.append(fields[i][j].getIndex() + "");
+			}
+			sb.append("\n");
+		}
+		return sb.toString();
+	}
+
+	@Override
+	public String toRowColString() {
+		StringBuilder sb = new StringBuilder();
+		for (int i = 0; i < rows; i++) {
+			for (int j = 0; j < columns; j++) {
+				sb.append(fields[i][j].getPosition().getRow() + ":"
+						+ fields[i][j].getPosition().getCol() + "");
+			}
+			sb.append("\n");
+		}
+		return sb.toString();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 155:163, 181:189.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.